src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations.FlowAnalysis;
+using static Microsoft.CodeQuality.Analyzers.MicrosoftCodeQualityAnalyzersResources;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations
+{
+    public abstract partial class AvoidMultipleEnumerations : DiagnosticAnalyzer
+    {
+        private const string RuleId = "CA1851";
+
+        private static readonly DiagnosticDescriptor MultipleEnumerableDescriptor = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(AvoidMultipleEnumerationsTitle)),
+            CreateLocalizableResourceString(nameof(AvoidMultipleEnumerationsMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.Disabled,
+            description: null,
+            isPortedFxCopRule: false,
+            isDataflowRule: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(MultipleEnumerableDescriptor);
+
+        /// <summary>
+        /// Additional types that has the ability to defer enumeration.
+        /// </summary>
+        private static readonly ImmutableArray<string> s_additionalDeferredTypes =
+            ImmutableArray.Create(WellKnownTypeNames.SystemLinqIOrderedEnumerable1);
+
+        /// <summary>
+        /// All the immutable collections that have a conversion method from IEnumerable.
+        /// </summary>
+        private static readonly ImmutableArray<(string typeName, string methodName)> s_immutableCollectionsTypeNamesAndConvensionMethods = ImmutableArray.Create(
+            ("System.Collections.Immutable.ImmutableArray", "ToImmutableArray"),
```

Comment:
Can we use `WellKnownTypeNames` here instead of hardcoding the fully qualified names?

Reply:
It is a static class : )
https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray?view=net-5.0